beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs

```diff
@@ -0,0 +1,472 @@
+//! Provides a mechanism which queues work for later processing.
+//!
+//! When the `beacon_processor::Worker` imports a block that is acceptably early (i.e., within the
+//! gossip propagation tolerance) it will send it to this queue where it will be placed in a
+//! `DelayQueue` until the slot arrives. Once the block has been determined to be ready, it will be
+//! sent back out on a channel to be processed by the `BeaconProcessor` again.
+//!
+//! There is the edge-case where the slot arrives before this queue manages to process it. In that
+//! case, the block will be sent off for immediate processing (skipping the `DelayQueue`).
+//!
+//! Aggregated and unaggregated attestations that failed verification due to referencing an unknown
+//! block will be re-queued until their block is imported, or until they expire.
+use super::MAX_SCHEDULED_WORK_QUEUE_LEN;
+use beacon_chain::{BeaconChainTypes, GossipVerifiedBlock, MAXIMUM_GOSSIP_CLOCK_DISPARITY};
+use eth2_libp2p::{MessageId, PeerId};
+use fnv::FnvHashMap;
+use futures::task::Poll;
+use futures::{Stream, StreamExt};
+use slog::{crit, debug, error, Logger};
+use slot_clock::SlotClock;
+use std::collections::{HashMap, HashSet};
+use std::pin::Pin;
+use std::task::Context;
+use std::time::Duration;
+use task_executor::TaskExecutor;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::error::Error as TimeError;
+use tokio_util::time::delay_queue::{DelayQueue, Key as DelayKey};
+use types::{Attestation, EthSpec, Hash256, SignedAggregateAndProof, SubnetId};
+
+const TASK_NAME: &str = "beacon_processor_reprocess_queue";
+
+/// Queue blocks for re-processing with an `ADDITIONAL_QUEUED_BLOCK_DELAY` after the slot starts.
+/// This is to account for any slight drift in the system clock.
+const ADDITIONAL_QUEUED_BLOCK_DELAY: Duration = Duration::from_millis(5);
+
+/// For how long to queue aggregated and unaggregated attestations for re-processing.
+pub const QUEUED_ATTESTATION_DELAY: Duration = Duration::from_secs(30);
+
+/// Set an arbitrary upper-bound on the number of queued blocks to avoid DoS attacks. The fact that
+/// we signature-verify blocks before putting them in the queue *should* protect against this, but
+/// it's nice to have extra protection.
+const MAXIMUM_QUEUED_BLOCKS: usize = 16;
+
+/// How many attestations we keep before new ones get dropped.
+const MAXIMUM_QUEUED_ATTESTATIONS: usize = 1_024;
```

Comment:
@dapplion astutely pointed out that this value may be too low if a large percentage of validators attest early. Assuming 250k validators, that's 7.8k per slot. We could set the max capacity to 8192 to give us a bit of headroom. What do you think @paulhauner, @divagant-martian?

Reply:
I agree. I also expected this to be a bit small but was planning to just run it and see what happens whilst it's in `unstable`.

That being said, I think 8,192 is a better starting point.

Reply:
Why not compute these numbers on start-up depending on the preset constants?

Reply:
> Why not compute these numbers on start-up depending on the preset constants?

I'm not sure I understand. Are you saying to compute it based on the current validator count?

I'm hesitant to over-think this, I'm keen to get it onto our staging infra, monitor it on Grafana and then start tweaking. The worst case is that the queue fills up and we fall back to the same behaviour as before this PR.